### PR TITLE
feat: allow worker to manage single GPUs

### PR DESCRIPTION
**Pull request by sshaman1101**
added:
- GPU Tuner collects and keep info about available GPUs and related DRI devices.
- DRI device representation, related /sys/class parsers for device discovering, PCI bus ID extractor (inspired by rocm-smi).
- Tuner now can bind devices into the container by their IDs.
- `lsgpu` tool for GPU fast-checking on a various systems.

changed:
- Proto's GPUDevice completely updated, now it can correctly describe both Nvidia and Radeon devices.
- OpenCL GPU discoverer now collects required minimum info about the device. This info is used to ensure that required device capabilities is present by user's hardware.
- Worker's GPU capabilities now presents by GPU tuners, not `Hardware.Info`.

### CODE REVIEW

File: cmd/lsgpu/main.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	"os"
+
+	"github.com/sonm-io/core/insonmnia/miner/gpu"
+)
+
+var appVersion string
+
+func main() {
+	fmt.Printf("sonm lspgu %s\r\n", appVersion)
+	cards, err := gpu.CollectDriCardDevices()
```
Comment (antmat) [inline cmd/lsgpu/main.go]: maybe CollectDRICardDevices ?

File: insonmnia/miner/gpu/nvidia_tuner.go [hunk 2]
```diff
@@ -58,26 +143,37 @@ func newNvidiaTuner(ctx context.Context, opts ...Option) (Tuner, error) {
 
 	defer func() { nvidia.Shutdown() }()
 
-	log.G(ctx).Info("NVIDIA GPU supported by the host. Discovering GPU devices")
```
Comment (antmat) [inline insonmnia/miner/gpu/nvidia_tuner.go]: restore

Comment (antmat) [reply insonmnia/miner/gpu/nvidia_tuner.go]: better to prefer comma instead of dot

Comment (sshaman1101) [reply insonmnia/miner/gpu/nvidia_tuner.go]: [Removed!](https://github.com/sonm-io/core/pull/491/commits/6688596f41d7ba69e1c67a40862d0c5a09321315?diff=split#diff-f6057e6d60f924533a29826beee17231R146)

File: proto/capabilities.proto [hunk 3]
```diff
@@ -41,21 +41,24 @@ message RAMDevice {
...
+    // ID returns unique device ID on workers machine,
+    // typically PCI bus ID
+    string ID = 1;
+
+    // VendorID returns an unique device vendor identifier
+    uint64 vendorID = 2;
+    // VendorName returns GPU vendor name.
     string vendorName = 3;
-    // Total maximum memory size the device can hold.
-    uint64 maxMemorySize = 4;
```
Comment (3Hren) [inline proto/capabilities.proto]: Don't we need this anymore?

File: insonmnia/hardware/hardware.go [hunk 4]
```diff
@@ -49,7 +39,7 @@ type HardwareInfo interface {
 	Memory() (*mem.VirtualMemoryStat, error)
 
 	// GPU returns information about GPU devices on the machine.
-	GPU() ([]gpu.Device, error)
+	// GPU() ([]*pb.GPUDevice, error)
```
Comment (3Hren) [inline insonmnia/hardware/hardware.go]: UDOLI

File: insonmnia/hardware/hardware.go [hunk 5]
```diff
@@ -3,15 +3,15 @@ package hardware
 import (
 	"github.com/shirou/gopsutil/mem"
 	"github.com/sonm-io/core/insonmnia/hardware/cpu"
-	"github.com/sonm-io/core/insonmnia/hardware/gpu"
+	pb "github.com/sonm-io/core/proto"
```
Comment (3Hren) [inline insonmnia/hardware/hardware.go]: No pb please

File: insonmnia/miner/gpu/nvidia_tuner.go [hunk 6]
```diff
@@ -120,3 +216,23 @@ func newNvidiaTuner(ctx context.Context, opts ...Option) (Tuner, error) {
 
 	return &ovs, nil
 }
+
+func makeVolumeMount(src, dst, name string) mount.Mount {
```
Comment (3Hren) [inline insonmnia/miner/gpu/nvidia_tuner.go]: `newVolumeMount`

File: insonmnia/miner/gpu/tuner_other.go [hunk 7]
```diff
@@ -8,7 +8,7 @@ import (
 
 // override all Tuner implementation with NilTuner if we building without GPU support
 
-func newRadeonTuner(opts ...Option) (Tuner, error) {
+func newRadeonTuner(_ context.Context, opts ...Option) (Tuner, error) {
```
Comment (3Hren) [inline insonmnia/miner/gpu/tuner_other.go]: Why?

File: insonmnia/miner/overseer.go [hunk 8]
```diff
@@ -62,8 +63,19 @@ func (d *Description) Mounts(source string) []volume.Mount {
...
 
-func (d *Description) GPU() bool {
-	return d.GPURequired
+func (d *Description) IsGPURequired() bool {
+	_, ok := d.Env["GPU"]
+	return ok
+}
+
+func (d *Description) GpuDeviceIDs() []gpu.GPUID {
+	// warn: hack just for testing, allows to set GPU device via task's ENV params
```
Comment (3Hren) [inline insonmnia/miner/overseer.go]: No hack.

**Pull request merged**